From 71dd0ef8b8d35a41e07f05c9b2c3e612dc52a57c Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Wed, 8 Jun 2022 06:46:02 +0000
Subject: [PATCH 2658/4585] [Huawei] mm/sharepool: Don't check the DVPP address
 space range before merging

Offering: HULK
ascend inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I5DS9S

-------------------------------------------------

The user doesn't care about the start address of the dvpp range, what
is mattered is that the virtual space tagged DVPP located at in a 16G
range. So we can safely drop the dvpp address space as long as it's
empty during merging process.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 23 ++++++++++++++++-------
 1 file changed, 16 insertions(+), 7 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 7e04b51dc985..59db0e27880b 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -241,16 +241,25 @@ static int sp_mapping_group_setup(struct mm_struct *mm, struct sp_group *spg)
 	struct sp_group *local = master->local;
 
 	if (!list_empty(&spg->procs) && !(spg->flag & SPG_FLAG_NON_DVPP)) {
-		if (!can_mappings_merge(local->dvpp, spg->dvpp)) {
-			pr_info_ratelimited("address space conflict, id=%d\n", spg->id);
-			return -EINVAL;
-		}
+		/*
+		 * Don't return an error when the mappings' address range conflict.
+		 * As long as the mapping is unused, we can drop the empty mapping.
+		 * This may change the address range for the task or group implicitly,
+		 * give a warn for it.
+		 */
+		bool is_conflict = !can_mappings_merge(local->dvpp, spg->dvpp);
 
-		if (is_mapping_empty(local->dvpp))
+		if (is_mapping_empty(local->dvpp)) {
 			sp_mapping_merge(spg->dvpp, local->dvpp);
-		else if (is_mapping_empty(spg->dvpp))
+			if (is_conflict)
+				pr_warn_ratelimited("task address space conflict, spg_id=%d\n",
+						spg->id);
+		} else if (is_mapping_empty(spg->dvpp)) {
 			sp_mapping_merge(local->dvpp, spg->dvpp);
-		else {
+			if (is_conflict)
+				pr_warn_ratelimited("group address space conflict, spg_id=%d\n",
+						spg->id);
+		} else {
 			pr_info_ratelimited("Duplicate address space, id=%d\n", spg->id);
 			return -EINVAL;
 		}
-- 
2.17.1

